xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-default/src/main/java/org/xwiki/mentions/internal/DefaultMentionsEventExecutor.java

```diff
@@ -0,0 +1,159 @@
+ * First, when {@link DefaultMentionsEventExecutor#execute(DocumentReference, DocumentReference, String)} is called,
+ * a {@link MentionsData} is added to a queue.
+ * Then, a pool of {@link MentionsConsumer} workers is consuming the queue and delegate the actual analyis to
+ * {@link MentionsDataConsumer#consume(MentionsData)} which deals with actual implementation of the user mentions 
+ * analysis.  
+ *
+ * @version $Id$
+ * @since 12.6RC1
+ */
+@Component
+@Singleton
+public class DefaultMentionsEventExecutor implements MentionsEventExecutor, Initializable, Disposable
+{
+    private ThreadPoolExecutor executor;
+
+    private BlockingQueue<MentionsData> queue;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private MentionsBlockingQueueProvider blockingQueueProvider;
+
+    @Inject
+    private MentionsThreadPoolProvider threadPoolProvider;
+
+    @Inject
+    private MentionsDataConsumer dataConsumer;
+
+    @Inject
+    private JMXBeanRegistration jmxRegistration;
+
+    @Override
+    public void initialize()
+    {
+        this.executor = this.threadPoolProvider.initializePool();
+        this.queue = this.blockingQueueProvider.initBlockingQueue();
+        IntStream
+            .range(0, this.threadPoolProvider.getPoolSize())
+            .forEach(i -> this.executor.execute(new MentionsConsumer()));
+        JMXMentionsMBean mbean = new JMXMentions(this.queue);
+        this.jmxRegistration.registerMBean(mbean, "name=mentions");
+    }
+
+    @Override
+    public void dispose()
+    {
+        this.jmxRegistration.unregisterMBean("name=mentions");
+    }
+
+    @Override
+    public void execute(DocumentReference documentReference, DocumentReference authorReference, String version)
+    {
+        try {
+            this.queue.put(new MentionsData()
+                               .setDocumentReference(documentReference.toString())
+                               .setAuthorReference(authorReference.toString())
+                               .setVersion(version)
+                               .setWikiId(documentReference.getWikiReference().getName())
+            );
+        } catch (InterruptedException e) {
+            this.logger
+                .warn("Error while adding a task to the mentions analysis queue. Cause [{}]", getRootCauseMessage(e));
+        }
+    }
+
+    @Override
+    public long getQueueSize()
+    {
+        return this.queue.size();
+    }
+
+    /**
+     * Consumer of the mentions analysis task queue.
+     */
+    public class MentionsConsumer implements Runnable
+    {
+        @Override
+        public void run()
+        {
+            while (true) {
+                runOnce();
+            }
+        }
+
+        /**
+         * Consume a single queue task.
+         */
+        public void runOnce()
+        {
+            MentionsData data = null;
+            try {
+                data = DefaultMentionsEventExecutor.this.queue.take();
```

Comment:
Not sure about this: there's always a data to take? Since it's running indefinitely isn't there some cases where the data is null?

Reply:
That's why we are using a BlockingQueue here, the take method is waiting passively if no element is available.

> Retrieves and removes the head of this queue, waiting if necessary until an element becomes available.
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingQueue.html#take--

Reply:
Makes sense. Thanks

Reply:
Just for info, the name of the method `runOnce` is a bit misleading then :) You could put everything in the while loop of run method.

Reply:
It just a convenient way to isolate an iteration. That's mainly for testing purpose.

Reply:
Ok then.